### ISSUE TITLE
`inbox` sometimes is nil

### ISSUE THREAD

**Issue by davidpelaez** [soveran/syro#19]
Hi.

This is very strange situation and I don't know for if you can replicate it, but here's my best try after quite some debugging. Sometimes `inbox` is `nil` but `self.inbox` has the actual expected values. In some blocks I don't write self, in other I do. I cannot understand why the difference.

Here's a file I'm have for reference, I put a comment where the `nil` problem is visible:

``` ruby
module HTTP
  WarehouseEndpoints = Syro.new(ERPDeck) do

    capture :warehouse_id and validate_warehouse_id

    on('lines') do

      get { json data: JSONDeck::LINES[inbox[:country_id]].to_h }

      capture :line_id and validate_line_id

     # Everything normal until here, inbox is a valid hash indeed. No need to do `self.inbox`



        on("products") do

          # HERE `inbox` is `nil` but `self.inbox` is the expected value. Why?

          setup_pagination
          dataset = Datasets::Products::in_warehouse inbox[:warehouse_id]
          defaults = { order_by: :id, order: :desc, page_size: 50, page: 1}
          inbox = defaults.merge(inbox)
          #binding.pry

          # Set SQL order
          if inbox[:order].to_sym == :desc
            dataset.order Seq.desc(inbox[:order_by])
          else
            dataset.order Seq.asc(inbox[:order_by])
          end

          get do
            products = dataset.where(line_id: inbox[:line_id])
            products = products.paginate(inbox[:page], inbox[:page_size]).all
            json data: strip(products.map_to_h)
          end

          on(:product_id) do
            product = dataset[id: inbox[:product_id]]
            get { json data: product.to_hash.strip }
          end
        end
      # }

      binding.pry


    end
  end
end

```

Is this because of a nested `on` block? I cannot think of another reason. Maybe the binding when yielding to a nested block affects the call to `inbox`?

Please let me know if it's an error on my side, I cannot find a reason as to why it's not working nor did I see anything in Syro's code that made me consider there was a problem.

Thanks

**Comment by soveran** [comment]
@davidpelaez Can you show me `ERPDeck`?

**Comment by davidpelaez** [comment]
Sure, I removed some small sensitive parts. It's totally WIP, so please excuse any nasty code:

``` ruby
class ERPDeck < Syro::Deck

  COUNTRIES = ["XY"]

  WAREHOUSES = {
    "CO": ["A", "b", "C"]
  }.to_mash

  LINES = {
    "CO": {
      "01" => "X",
      "02" => "Y",
      "03" => "Z"
    }
  }.to_mash

  def json(model)
    res[Rack::CONTENT_TYPE] = "application/json; charset=utf-8"
    res.write(model.to_json)
  end

  def not_found_error(msg='No se encontró el recurso solicitado')
    res.status = 404
    msg = { error: 'not_found', message: msg }
    json(msg) and finish_request
  end

  def finish_request
    halt(res.finish)
  end

  def invalid_data_error
    res.status = 424
    msg = { error: 'unprocessable_entity', message: "Atributos inválidos" }
    json(msg)
  end

  def parse_body
    # TODO rescue potential error
    inbox.merge! Oj.load(req.body.read)
  end

  def parse_params
      inbox.merge! Hashie.symbolize_keys(req.params)
  end

  def validate_country_id
    unless COUNTRIES.include? inbox[:country_id]
      not_found_error("Código de país inválido")
    end
  end

  def validate_warehouse_id
    country = inbox[:country_id]
    unless WAREHOUSES[country].include? inbox[:warehouse_id]
      not_found_error("Código de bodega inválido")
    end
  end

  def validate_line_id
    country = inbox[:country_id]
    unless LINES[country].include? inbox[:line_id]
      not_found_error("Código de línea inválido")
    end
  end

  def apply_collection_constraints(dataset)
    paginate_collection filter_collection(sort_collection(dataset))
  end

  def sort_collection(dataset)
    inbox.apply_defaults({order_by: :id, order: :desc})
    inbox[:order_by] = inbox[:order_by].to_sym
    # Set the order to a valid Sequel expression
    if inbox[:order].to_sym == :desc
      inbox[:order] = Sequel.desc(inbox[:order_by])
    else
      inbox[:order] = Sequel.asc(inbox[:order_by])
    end
    return dataset.order(inbox[:order])
  end

  def paginate_collection(dataset)
    inbox.apply_defaults( page_size: 50, page: 1)
    # Ensure that pagination params are real numbers
   inbox[:page] = inbox[:page].to_i
   inbox[:page_size] = inbox[:page_size].to_i
   return dataset.paginate(inbox[:page], inbox[:page_size])
  end

  def filter_collection(dataset)
    dataset # NO OP in the meantime
  end

  def strip(o)
    o.flatten.map(&:strip)
  end

end
```

**Comment by soveran** [comment]
In the app, at some point you do this:

``` ruby
inbox = defaults.merge(inbox)
```

Up to that point, `inbox` is a method that refers to an instance variable, but after that it's just a local variable. Can you check what happens if you remove that line?

**Comment by davidpelaez** [comment]
I had already removed that part and after testing again it seems to work. I think it was indeed the fact that it pointed to a local variable and there was some confusion. I preferred using `inbox.merge! defaults.merge(inbox)` and it works. I also noticed that if I indeed want to assign `inbox` I have to do it directly to `@syro_inbox`. It was indeed a bug on my code, thank you for taking the time to look at it. Also thanks for creating Syro, I'm loving the simplicity and productivity it brings :)
